### ISSUE TITLE
Odd mapping generated for FishTankMapper

### ISSUE THREAD

**Issue by gunnarmorling** [mapstruct/mapstruct#1060]
Something looks weird there. Why does it generate a `fishTankToFishDto()` method? Two questions I see really:

* why does it try to map fish tank to fish DTO (the elements are on different levels in the tree)?
* why is this an update method instead of a regular mapping method with a return value?

**Comment by sjaakd** [comment]
> why does it try to map fish tank to fish DTO (the elements are on different levels in the tree)?

Nested mappings were never limited to a symmetrical tree. It works exactly as before.

>why is this an update method instead of a regular mapping method with a return value?

Checkout: NestedTargetPropertiesTest#shouldMapNestedComposedTarget. I need update methods because the mapping for the first source parameter would influence the mapping for the next source parameter.

The generated code works exactly as before (even before the auto-mapping). I only added the FishTankMapper test case to show a problem in my initial solution, which was shown by tests of @snowe2010. See issue #1011. I only make use of the auto-mapping do simplify the code and to solve the problem in #1011.

For more info, see #1057 for more info.

**Comment by gunnarmorling** [comment]
> Nested mappings were never limited to a symmetrical tree.

Yes, ok, but here the tree is symmetrical, right? So I'm wondering why we don't create a mapping method for each level in the tree. It works as I'd expect for `WaterPlant(Dto)`, it's not clear to me why it's different for `Fish`.

> I need update methods because the mapping for the first source parameter would influence the mapping for the next source parameter.

But we only have a single source parameter in the fish tank case, no?

> The generated code works exactly as before (even before the auto-mapping)

Hum, but the automatic mapping of fish and water plant without any user-provided method is only done since the introduction of auto-mapping, right?

**Comment by sjaakd** [comment]
> Yes, ok, but here the tree is symmetrical, right? So I'm wondering why we don't create a mapping method for each level in the tree. It works as I'd expect for WaterPlant(Dto), it's not clear to me why it's different for Fish.

That's exactly why I wrote #1057 and send you guys mails. I think we never considered the combination nested source  / nested target properties icw auto-mapping. I just found out while solving an entirely different problem and was triggered by some remarks @agudian made about the eclipse plugin. 

You could indeed claim that the (defined by @Mapping) symmetrical mappings are a subset of the complete set of mappings. We have to recognize them in a bean mapping and they should be handled separately. I don't think this is easy.

> But we only have a single source parameter in the fish tank case, no?

True.. But what do we gain by adding complexity to our logic (being an update method to start with, one parameter, two parameters and perhaps later the presence of symmetrical mappings) to decide what kind method to use. The generated methods are private right? In efficiency they are (most likely) not so much worse than create methods. You could also make a plea for the case that simplicity (always using the same style) benefits the readability of generated code as well. 


> Hum, but the automatic mapping of fish and water plant without any user-provided method is only done since the introduction of auto-mapping, right?

Precisely.  See my answer just above. 

WDYT: perhaps we should mark this as duplicate of #1057 ?

**Comment by filiphr** [comment]
I have been thinking about this one a lot. I have some ideas how it can be solved, but I got stuck at one point.

My main idea was not to force updated generated mapping for the nested target properties. If we do this then we will generate `private FishTankDto fishTankToFishDto(FishTank fishTank)`. @gunnarmorling the generated mapping for `Fish` is different that `WaterPlant(Dto)` because we have a nested target property.

I am stuck with composite mappings.  For example we have the following: 

```java
@Mappings({
    @Mapping(target = "type", ignore = true),
    @Mapping(target = "name", source = "chartEntry2.chartName"),
    @Mapping(target = "song.title", source = "chartEntry1.songTitle" ),
    @Mapping(target = "song.artist.name", source = "chartEntry1.artistName" ),
    @Mapping(target = "song.artist.label.studio.name", source = "chartEntry1.recordedAt"),
    @Mapping(target = "song.artist.label.studio.city", source = "chartEntry1.city" ),
    @Mapping(target = "song.positions", source = "chartEntry2.position" )
})
public abstract Chart map(ChartEntry chartEntry1, ChartEntry chartEntry2);
```
from our `ChartEntryToArtist` mapper. Here we have multiple parameters that are used in the mapping. The generated code of this method looks like this:

```java
@Override
public Chart map(ChartEntry chartEntry1, ChartEntry chartEntry2) {
    if ( chartEntry1 == null && chartEntry2 == null ) {
        return null;
    }

     Chart chart = new Chart();

    if ( chartEntry1 != null ) {
        if ( chart.getSong() == null ) {
            chart.setSong( new Song() );
        }
        chartEntryToSong1( chartEntry1, chart.getSong() );
    }
    if ( chartEntry2 != null ) {
        if ( chart.getSong() == null ) {
            chart.setSong( new Song() );
        }
        chartEntryToSong2( chartEntry2, chart.getSong() );
        chart.setName( chartEntry2.getChartName() );
    }
     return chart;
}
```

Here creating the _update_ methods is OK because we create property mappings per source parameter. Here we need to "share" the song between the 2 source parameters. However, my idea is to continue creating mappings for each level in the nested target properties and send the needed parameters down to those mappings. My idea for this is to generate something like:

```java
@Override
public Chart map(ChartEntry chartEntry1, ChartEntry chartEntry2) {
    if ( chartEntry1 == null && chartEntry2 == null ) {
        return null;
    }

     Chart chart = new Chart();

    if ( chartEntry2 != null ) {
        chart.setName( chartEntry2.getChartName() );
    }

    chart.setSong( mapSongFromChartEntry1ChartEntry2(chartEntry1, chartEntry2) );
     return chart;
}

//The annotations are only added to display how the MappingOptions will look like
@Mapping(target = "title", source = "chartEntry1.songTitle" ),
@Mapping(target = "artist.name", source = "chartEntry1.artistName" ),
@Mapping(target = "artist.label.studio.name", source = "chartEntry1.recordedAt"),
@Mapping(target = "artist.label.studio.city", source = "chartEntry1.city" ),
private Song mapSongFromChartEntry1ChartEntry2(ChartEntry chartEntry1, ChartEntry chartEntry2) {
    //One can even contest that we do not need to do a null check here as it is an internal method and we couldn't have send 2 nulls
    if ( chartEntry1 == null && chartEntry2 == null) {
        return null;
    }

    Song song = new Song();

    if ( chartEntry1 != null ) {
        song.setTitle( chartEntry1.getSongTitle() );
    }

    song.setArtist( mapArtistFromChartEntry1ChartEntry2( chartEntry1, chartEntry2 );

    return song;
}

//... and we continue going down and passing the mappings until we are done with the nested target properties
```

I am stuck with the current implementation of the `BeanMappingMethod` where we do mappings only per source property and we can't group them together. I have been trying to think of an alternative way to do it, but we need both the current implementation and the grouped. By trying to achieve both I am blocked by the cycle detection, I am not sure if I am on the right path or not.

**Comment by sjaakd** [comment]
> However, my idea is to continue creating mappings for each level in the nested target properties and send the needed parameters down to those mappings

@filiphr : we discussed that earlier :smile: , I believe that this is something I mentioned. However, I dismissed it. Method selection only works on single sources. Here you require method selection of a method with 2 source parameters. That will spin out of control. So for this particular scenario its ok to use update methods. That's what I meant in our meeting on you need to have some logic to decide update vs create. 
1. Having 2 parameters is an update method. 
2. Being an update method means creating an update method
3. ..
 I consider this create / update a matter of beautifying the solution.

Having said that, this is not the main problem to focus on. I think that we should start out symmetric and diverge as soon as it becomes asymmetric (passing the source parameter only at that point in the object graph) is the main problem. So the current solution is OK at a point in time the source / target is not symmetric anymore. The Fishtank case is symmetric. 

Hope I make sense

**Comment by filiphr** [comment]
I think that we can close this as with #1071 we create update methods only for composite (more than one source parameter) methods.

WDYT @gunnarmorling?

**Comment by sjaakd** [comment]
@filiphr are we going to open up a new issue to take my points above? Or do we address those as part of #1060.?

**Comment by filiphr** [comment]
@sjaakd I think that we can create a new one or move the discussion to #1057 (are they connected?)

**Comment by sjaakd** [comment]
Look at the original text above:

> why is this an update method instead of a regular mapping method with a return value?

This is addressed in your PR.


> why does it try to map fish tank to fish DTO (the elements are on different levels in the tree)?

This is not yet addressed. But I think this is the main problem (and what #1057 actually describes). See also my own comment above. I think we should close one of the issues since they address the same thing, so lets continue in #1057 with the remainder and close this one.

**Comment by gunnarmorling** [comment]
@sjaakd @filiphr This seems to be addressed with PR #1057, right? I can't see any update methods being used any more in the fish tank example.

**Comment by filiphr** [comment]
@gunnarmorling This actually was resolved with PR #1071

**Comment by gunnarmorling** [comment]
Ok, so I'm going to close this one then. If there is anything from the discussions above which should be kept, please open a separate issue (nothing seems left from a quick glance, though). Thanks!
